### PR TITLE
schema-gen: YAML-driven proto-to-schema codegen + migrate all resources to it

### PR DESCRIPTION
**Pull request by gene-redpanda**
Here's the final body with the test plan section removed:

  ## Summary
  
  Introduces `schemagen`, a YAML-driven codegen pipeline that walks cloudv2
  protos and emits the Terraform framework schema, Go model structs,
  Flatten/Expand conversion code, and protovalidate-backed validators for each
  resource. Every resource and datasource is migrated onto it. Hand-written code
  shrinks to CRUD glue plus per-resource hooks; schema/model/conversion are
  generated and pinned by golden files.

  Also lands a three-tier test stack (unit / bufconn integration / live-acc),
  on-disk OAuth token caching, upstream dependency pinning, and a batch of
  correctness fixes and doc/example refreshes.

  ## Codegen pipeline
  
  - `internal/apidesc` — OpenAPI description loader (Encode/Lookup/Filter)
  - `internal/schemagen` — proto walker + YAML-directive merge; emits framework
    schema, model structs, Flatten/Expand, and validators
  - `cmd/schemagen`, `cmd/enumgen`, `cmd/apidesc-import` — codegen binaries
  - Per-resource config in `redpanda/resources/<r>/schema.yaml`; central
    `//go:generate` registry in `schemagen.go`; global config in `codegen.yaml`
  - `internal/bufdeps` pins cloudv2/console SHAs + buf-gen BSR commits in
    `internal/buf_dependencies.yaml`; `task generate:bump-deps` realigns
    everything against a freshly pulled upstream
  - Retires `pkg/generate/modelgen/`

  Generated output ships in dedicated `chore(generated)` commits placed after
  each feature commit so reviewers can skim and skip them.

  ## Resource migration

  Migrated to schemagen: resource_group, network, cluster, user, role, acl,
  schema_registry_acl, topic, pipeline, secret, shadow_link, serverless_cluster,
  serverless_private_link, and the read-only datasources (region, regions,
  serverless_regions, throughput_tiers). roleassignment stays hand-written (Kafka
  API surface, no dataplane proto to walk) but gains a principal-prefix validator
  and composite import parser.

  Per-resource hooks preserve the behaviors the generated round-trip can't derive:
  sensitive-field preservation across masked reads (shadow_link client_options,
  pipeline/service_account secrets, secret material), customer_managed_resources
  shape preservation, cloud_provider_tags reconciliation, and various
  RequiresReplace rules.

  ## New / changed surface
  
  - **New** `redpanda_service_account` control-plane resource
  - `redpanda_schema` datasource is now functional (was a "not yet implemented"
    stub); bearer + Basic auth supported
  - `redpanda_serverless_cluster` gains an `allow_deletion` destroy gate
    (previously the only billable resource with no guard)
  - `redpanda_serverless_private_link`: `cloud_provider_config` deprecated and
    replaced by `aws_config`; old shape now fails the plan with a templated
    migrate (`state rm` / `import`) message instead of "Unsupported argument"
  - `redpanda_schema` import ID separator changed `:` → `,` (colons are valid in
    Avro/Protobuf subject names)

  ## Testing
  
  - **Integration tier** (`internal/testutil`): bufconn-backed gRPC server with
    per-service in-memory fakes, step builders (Create/Noop/RequiresReplace/
    UpdateLeaf/ImportRoundTrip/ErrorPath), plus acc + upgrade harnesses. Each
    resource gets `integration_*_test.go` and `upgrade_*_test.go`.
  - **Live-acc tier**: runner restructure, build-tag gating (replaces `RUN_*`
    env vars), BYOVPC sweepers. CI labels: byoc split out of `ci-ready` into
    `ci-ready-byoc`; labels otherwise unchanged.
  - Fakes uphold a production-realism contract (mask sensitive material, inject
    server-managed fields) so the integration tier exercises the same provider
    hooks the live backend does.

  ## Auth

  - OAuth tokens cached on disk at `<UserCacheDir>/redpanda/provider/creds-cache.json`
    via a layered `oauth2.TokenSource`; reused across `terraform` invocations to
    stay under the Auth0 daily quota. Disable with `REDPANDA_TOKEN_CACHE_DISABLE=1`.
  - TokenSource plumbed end-to-end through connpool, byoc, and every per-resource
    client; access tokens never logged.

  ## Other

  - Logging: per-poll/per-request mechanics → TRACE, retry chatter → DEBUG, new
    INFO create/update/delete milestones and wait-entry logs
  - Fixes: Avro canonicalization signature restore, schema-compatibility validator
    (no silent BACKWARD substitution), topic `redpanda.*` server-injected config
    strip, cluster plan-time enum validation restore on cloud_provider/connection_type
  - CI: BuildKite byoc split out of `ci-ready`, gotestsum→`go test` migration,
    lint-no-diff gate
  - Docs/examples refreshed (throughput tier value, provider version pin,
    sensitive vars in example HCL); README/TESTING/CLAUDE.md updated

### CODE REVIEW

**Comment by gene-redpanda** [comment]
non-serverless https://buildkite.com/redpanda/terraform-provider-redpanda/builds/1401
serverless https://buildkite.com/redpanda/terraform-provider-redpanda/builds/1405

File: internal/schemagen/validators.go [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+	Expr string
+
+	Imports []string
+
+	AttrType string
+
+	ReturnsSlice bool
+
+	Parameterized bool
+
```
Comment (andrewstucki) [inline internal/schemagen/validators.go]: non-blocking, but the spacing of this file seems real odd -- would you mind fixing it in a follow-up?

File: internal/schemagen/validator_descriptions.go [hunk 2]
```diff
@@ -0,0 +1,537 @@
+// Copyright 2023 Redpanda Data, Inc.
+//
+//
```
Comment (andrewstucki) [inline internal/schemagen/validator_descriptions.go:3]: same here -- wonder if it'd make sense to add in something like `license-validator` or whatever the thing I wrote that we leverage in the operator repo is called

File: internal/schemagen/proto_compiler.go [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+
+// CompileProtoFiles compiles every .proto file under cloudv2Root/proto/.../<protoPkg>
+// (plus any extraImportPaths) and returns the resolved file set. Files that
+// fail to compile (e.g. because of unresolvable imports) are skipped with a
+// log line — same forgiving behavior the original CompileProto used.
+func CompileProtoFiles(cloudv2Root, protoPkg string, extraImportPaths []string) (linker.Files, error) {
+	importPaths := []string{
+		filepath.Join(cloudv2Root, "proto", "public", "cloud"),
+		filepath.Join(cloudv2Root, "proto", "public", "common"),
+		filepath.Join(cloudv2Root, "tools", "proto", "wellknown"),
```
Comment (andrewstucki) [inline internal/schemagen/proto_compiler.go:57]: wonder if it'd make sense to not hard-code these, maybe move some of this to like a configuration file eventually? again, not blocking, but lots of this code seems like it could be somewhat useful as a generic protobuf --> terraform data model compiler...

File: internal/schemagen/generator.go [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+	"strings"
+	"text/template"
+)
+
+// LicenseHeader returns the standard license header for generated files.
+func LicenseHeader() string { return licenseHeader }
+
+const licenseHeader = `// Copyright 2023 Redpanda Data, Inc.
+//
+//
```
Comment (andrewstucki) [inline internal/schemagen/generator.go:32]: seems odd to have double lines, though it looks like this is generally the case in every file in this repo

File: internal/schemagen/flatten_expand_plan.go [hunk 5]
```diff
@@ -0,0 +1,1749 @@
...
+		return ".GetValue()"
+	}
+	return ""
+}
+
+// renderScalarHeuristicFlatten emits the three-arm FlattenStmt used when a TF
+// attribute is Optional-only but the proto field carries no presence: prefer
+// the proto value when non-zero, fall back to prev (preserving user
+// null-vs-zero intent), then null. Used for every scalar AttrType.
+func renderScalarHeuristicFlatten(conv *FieldConversion, sh scalarShape) string {
```
Comment (andrewstucki) [inline internal/schemagen/flatten_expand_plan.go:65]: If I'm understanding this correctly -- if a TF attribute it marked as "optional" but the proto value is a raw scalar, we can't distinguish between the 0-value and a null-value from the proto side of things, so we fall back to the previously set value. Two questions:

1. Do we know of any fields where the zero value is actually meaningful?
2. If cloud has some defaulting behavior when something is unset/0, I'm assuming this could cause bugs if something was reset to unspecified?

Can we potentially flag this as a concern to the cloud team?

From a claude-helped audit -- seems like the two most concerning fields might be:


Field | Path | Type | Zero meaningful?
-- | -- | -- | --
HourOfDay | cluster.maintenance_window | int32 | Yes — 0 = midnight is valid
SkipDestroy | cluster.customer_managed_resources…cloud_storage | bool | Likely yes — false is the default

If I'm thinking through this correctly that means that the TF state wouldn't get updated on server-side updates despite a refresh if the server returns a 0 value. That also means that if the user really wanted to change the value back they'd have to temporarily modify it in terraform, apply, and then switch it back?

I'm honestly not sure if there's much to do here if these fields are actually optional other than potentially flag it to cloud.

**Review by andrewstucki — APPROVED** [review]
Only one real potential bug that jumped out to me from reading through the comments. There's some formatting oddities and we may need to raise a request to cloud for when they're publishing some protos to indicate whether some 0 value fields are actually optional or not, but seems to me like some might be buggy in the current implementation. That said, given the size of this, think this should be done likely as a follow-up.

File: internal/schemagen/validators.go (hunk 1 above)
Comment (gene-redpanda) [reply internal/schemagen/validators.go]: resolved

File: internal/schemagen/validator_descriptions.go (hunk 2 above)
Comment (gene-redpanda) [reply internal/schemagen/validator_descriptions.go:3]: Deferred for now because I want to add license-validator to avoid future license shenanigans.

File: internal/schemagen/proto_compiler.go (hunk 3 above)
Comment (gene-redpanda) [reply internal/schemagen/proto_compiler.go:57]: I deduped it. Yeah I'd agree, it'd be generally useful with a bit more effort. Its not 100% of the way there but it is definitely like 60-70% of the way.

File: internal/schemagen/generator.go (hunk 4 above)
Comment (gene-redpanda) [reply internal/schemagen/generator.go:32]: Deferred, same reason as the other license one. And I agree it's odd but I didn't want to get dinged with an even bigger PR 😂

File: internal/schemagen/flatten_expand_plan.go (hunk 5 above)
Comment (gene-redpanda) [reply internal/schemagen/flatten_expand_plan.go:65]: I've marked em computed and added some testing -- which flagged on this case and went green with the fixes -- so good spots and thank you. My expectation is that there are gonna be edge cases like this one that haven't been caught by the time we go live. I've tried to neutralize as many as possible with the additional testing layer and more thorough acceptance tests but with so many fields and so many permutations and combinations I doubt I've caught everything.
